infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,276 @@
+package org.corfudb.infrastructure.logreplication;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationEvent;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationState;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationStateType;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.IllegalTransitionException;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeEvent;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeState;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeStateType;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.sink.LogReplicationSinkEvent;
+import org.corfudb.runtime.LogReplication.LogReplicationSession;
+
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This class manages the tasks submitted by the runtime FSM and the replication FSM.
+ * Both the FSMs have their own instances of this class to which tasks are submitted and executed.
+ *
+ * The threads are shared by all the sessions traversing an FSM.
+ */
+@Slf4j
+public class FsmTaskManager {
+
+    private static ScheduledExecutorService runtimeWorker = null;
+    private static ScheduledExecutorService replicationWorker = null;
```

Comment:
minor: any reason for using a scheduled executor service for these 2 pools?  Can we use a simple executor service instead?

Reply:
replication worker thread pool will have scheduled tasks. Like the verification if apply phase has completed and for routing queue model to check if there is data.

runtime worker currently doesn't have any scheduled tasks. I can change that to be the simple executor